src/shared_.toit

```diff
@@ -0,0 +1,136 @@
+      // Drop "./" segments.
+      i += 2
+      continue
+    if at-separator and
+        bytes[i] == '.' and
+        bytes[i + 1] == '.' and
+        bytes[i + 2] == separator:
+      // Discard the previous segment (between the last two separators).
+      if separators.size < 2:
+        // We don't have a previous segment to discard.
+        if is-rooted:
+          // Just drop them if the path is absolute.
+          i += 3
+          // No need to update `at-separator`. It's still true.
+          continue
+        // Otherwise we have to copy them over.
+        bytes[target-index++] = bytes[i++]
+        bytes[target-index++] = bytes[i++]
+        bytes[target-index++] = bytes[i++]
+        // Move the separator.
+        // If we add a new segment that should be removed, it must be removed up
+        // to the current position. Anything before should not be touched.
+        // This happens when we have leading '..' that need to stay, like in `../abc/..`.
+        separators[0] = target-index - 1
+        // It's not a problem if 'target_index' is one after the '\0' (equal to
+        // 'bytes.size'), but it feels cleaner (and more resistant to future
+        // changes) if we fix it.
+        if target-index > path-size: target-index--
+        // No need to update `at-separator`. It's still true.
+        continue
+      // Still handling '..' here.
+      // Reset to the last '/'.
+      separators.resize (separators.size - 1)
+      target-index = separators.last + 1
+      i += 3
+      // No need to update `at-separator`. It's still true.
+      continue
+
+    if bytes[i] == separator:
+      separators.add target-index
+      at-separator = true
+    else:
+      at-separator = false
+
+    bytes[target-index++] = bytes[i++]
+
+  if target-index == volume-name-size:
+    if volume-name-size == 0: return "."
+    if volume-name-size > 2:
+      // Probably unreachable, but can't hurt to special case.
+      // Must be a UNC path.
+      // For example '//host/share'
+      return path[..volume-name-size]
+    // Must be a relatidrive path.
+    // For example `c:foo/..` which gets cleaned to `c:.`
+    bytes[target-index++] = '.'
+
+  // Drop trailing path separator unless it's the root path.
+  last-char-index := target-index - 1
+  if last-char-index > start-index and bytes[last-char-index] == separator:
+    target-index--
+
+  return bytes[..target-index].to-string
+
+/**
+Convert the $path to a relative path in relation to $base.
+
+If it is not possible to be relative, returns the absolute path of $path.
+
+The result is cleaned by $clean before being returned.
+*/
+to-relative_ path/string base/string --handle-different-root/bool=false -> string:
+  absolute-path := to-absolute path
+  absolute-base := to-absolute base
+
+  split-path := split absolute-path
+  split-base := split absolute-base
+
+  if handle-different-root and not split-path.is-empty and not split-base.is-empty:
```

Comment:
Maybe use `volume-name-size_` instead?

Should we say that `//host/share1/foo` is relative to `//host/share2/bar`?
If it is, then looking at the first segment is good enough.

Reply:
split will split a unc path to ['//host', 'share1', 'foo'], so it should already be doing that?

Reply:
But should it?
Or should `split` make `\\host\share1` a single entry in the returned list?

Reply:
Wait, you are right. \\host\share1\tmp and \\host\share2\tmp2 can not be relative to each other on windows (at least not in powershell). So we might as well let split do as you want.